feat(data-modeling): implement automatic relationship inference algorithm COMPASS-9776

gribnoysup opened the pull request:
This patch adds the logic for automatically discovering relationships between collections following the algorithm we outlined in the "Infer relationships" scope. This behavior is behind a feature flag until we add the toggle to disable the discovery to the schema generation UI, this is the next chunk of work for the project.

I added some collections to the `sample_airbnb` database in our teams "Cluster 0" cluster if you want to test it. If you create a data model for this namespace, it should identify relationships between `listingsAndReview`, `reviewers`, and `reviews` collections:

<img width="1138" height="558" alt="image" src="https://github.com/user-attachments/assets/46c88dea-8581-4715-a76c-d8409e42f941" />

Anemy commented on packages/data-service/src/data-service.ts:
```diff
@@ -1041,6 +1042,24 @@ class DataServiceImpl extends WithLogContext implements DataService {
+    executionOptions?: { fallbackReadPreference?: ReadPreferenceMode }
+  ) {
+    if (!options || !executionOptions) {
+      return options;
```
If options aren't set don't we still want to use the fallback read preference? 

Nice cleanup here and in indexes!

gribnoysup replied:
Oh, whoops, good catch! Let me add some coverage for this also

Anemy commented on packages/compass-data-modeling/src/store/analysis-process.ts:
```diff
@@ -161,18 +175,17 @@ export function startAnalysis(
       const collections = await Promise.all(
         namespaces.map(async (ns) => {
-          const sample: AggregationCursor<Document> = dataService.sampleCursor(
+          const sample = await dataService.sample(
```
Ah I see — we use all of the sample documents in the relationship check to see how many match each to ensure 1:1 .

gribnoysup replied:
When I was doing some performance testing for schema analysis, using cursor vs fetching first then passing the batch for analysis made virtually no meaningful difference in performance, so I'm confident this is not downgrading the performance here

addaleax commented on packages/compass-data-modeling/src/store/analysis-process.ts:
```diff
@@ -194,26 +207,71 @@ export function startAnalysis(
+                      logger.log.warn(
+                        mongoLogId(1_001_000_371),
+                        'DataModeling',
+                        'Failed when identifying relationship for the collection',
```
```suggestion
                        'Failure while identifying relationship for collection',
```

or
```suggestion
                        'Failed to identify relationship for collection',
```

addaleax commented on packages/compass-data-modeling/src/store/relationships.ts:
```diff
@@ -0,0 +1,215 @@
+    schema: MongoDBJSONSchema,
+    path: string[],
+    isArrayItem: boolean
+  ) => void,
```
I don't know if it's just me, but my first thought reading this function is "this callback could have been a generator". Obviously not really all that different in the end, but mentioning it in case you also feel that that would be more idiomatic

addaleax commented on packages/compass-data-modeling/src/store/relationships.ts:
```diff
@@ -0,0 +1,215 @@
+  }
+  const indexes = await dataService
+    .indexes(foreignNamespace, { full: false })
+    .catch(() => {
```
Just making sure – we're not logging anything because we know that the `DataService` method already will, correct?

addaleax commented on packages/compass-data-modeling/src/store/relationships.ts:
```diff
@@ -0,0 +1,215 @@
+              foreignNamespace,
+              {
+                _id: {
+                  $in: sampleDocs as any[], // driver wants this to be an ObjectId?
```
Yeah, it's trying to be helpful by assuming that `_id` is always an `ObjectId` unless you explicitly use something like `db.collection<{ _id: any }>('foo').count(...)`. It feels a bit overeager to me too but honestly for the typical application developer it's probably more helpful than unhelpful

gribnoysup replied:
Hmm, seems like specifically passing `any` as a type for a field is tripping up the generic logic for the Filter type in the driver ending up requiring `_id` to be `never`. I'll leave this as an assertion for now, but will revisit

gribnoysup commented on packages/compass-data-modeling/src/store/relationships.ts:
```diff
@@ -0,0 +1,215 @@
+ *
+ * @internal exported only for testing purposes
+ */
+export function traverseMongoDBJSONSchema(
```
They are not and for my purposes I'd like the types to be completely unwrapped and visited as single variants, I can sort of kind of do this on top of the method that was added, but at this point I'll have to add more or less the same amount of code to do that. We're planning to revisit traversal later, so that would probably a good point to figure out which one is the most generic method that can be shared

gribnoysup commented:
@addaleax @Anemy Thanks for the reviews! I'm rerunning some flaky tests, but otherwise I think I addressed all the feedback, do you want to take another look when you have a moment?

gribnoysup commented on packages/data-service/src/data-service.ts:
```diff
@@ -2211,26 +2237,43 @@ class DataServiceImpl extends WithLogContext implements DataService {
-    });
+    return indexes
+      .filter((index): index is IndexDescriptionInfo & { name: string } => {
+        return !!index.name;
```
I don't really know, that's the type we're getting directly from the driver, so I just assumed in some corner cases it might be possible and decided to account for that, can do a bit more digging 🙂
